### ISSUE TITLE
Adding getters to Span?

### ISSUE THREAD

**Issue by felixbarny** [opentracing/opentracing-java#57]
In this issue, I want to discuss adding getters to the Span interface. For example

``` java
Map<String, Object> getTags();
Object getTag(String key);
```

Getters for the Logs will probably be a bit more challenging though as there is no `LogData` interface in the api. 

The main reason for me is that it is currently impossible to write a custom reporter which is independent of a particular implementation.

Another reason as mentioned in #48:

> One use case where the client might want to access the tags the tracing library has created is when the lib parses the User-Agent header and the client wants to access the information whether the request was initiated from a mobile device.

**Comment by yurishkuro** [comment]
We don't know how implementations choose to deal with tags and logs. What if they are sending them out of process immediately without buffering in the Span object?  Or, more realistically, if a trace happens to take longer than a certain threshold, the implementation may decide to flush its data rather than waiting forever for the span to finish. By adding the getters that you suggest we'd be limiting the implementations in what they can do.

> The main reason for me is that it is currently impossible to write a custom reporter which is independent of a particular implementation.

OT API has no notion of reporters, so you're already in the realm of a particular implementation.

> when the lib parses the User-Agent header and the client wants to access the information whether the request was initiated from a mobile device.

Tracing libs do not parse headers, that is done by the RPC framework. For example, in TChannel the carrier (a map of key/values) that TChannel passes to the tracer only contains the tracing keys, none of the other headers. If the client code needs access to the headers, it should use the RPC framework for that.

**Comment by felixbarny** [comment]
> OT API has no notion of reporters, so you're already in the realm of a particular implementation.

I want to use OpenTracing and jaeger as the base for [stagemonitor](http://www.stagemonitor.org). I also want to write my own reporter classes instead of using Jaeger's `Reporter` interface. In my reporter classes I constantly have to cast the span to `com.uber.jaeger.Span`. Ideally, my reporters should not be aware of the fact that I'm using jaeger.

This change would also allow to write a library which can send any OpenTracing compliant Span implementation to Zipkin. Something that would otherwise require a lot of duplicated code between different implementations as this is basically a must for every implementation.

All in all I think this change would greatly simplify to switch to a different OpenTracing implementation.

**Comment by yurishkuro** [comment]
> Ideally, my reporters should not be aware of the fact that I'm using jaeger.

The concept of a reporter is an implementation detail of Jaeger, it's not a concept in the API, so I don't think it's fair to translate your expectations on the implementation into the API. If you want to work in the space of "generic" implementation, you can look at the abstract tracer in the `-impl-java8' module.

**Comment by mabn** [comment]
@yurishkuro These implementations are package protected. Is it intentional?

**Comment by michaelsembwever** [comment]
@felixbarny @yurishkuro 
 The reporter concept would be a welcome addition into the opentracing-java-impl codebase. The reporter approach has more merit than the current approach taken (for example in https://github.com/openzipkin/brave-opentracing )

**Comment by felixbarny** [comment]
Some more use cases where I need to access getters in my monitoring project stagemonitor which I'm currently refactoring to be built on top of OT:

- Only report spans which are slower than a (dynamic) threshold
- Include/exclude spans with a certain operation name. Note that the operation name can be set any time during the execution and can't be determined before creating the span.
- Increment timers based on the operation name
- There are many cases where I first put information in the tags which I then later need to retrieve again. Currently, there are about 50 usages of the getters in stagemonitor's code (excluding tests)

I know that I'm not the "typical audience" this API was designed for as I'm not an end user but a framework/library developer. At the moment I have to cast the Span at these points which makes stagemonitor quite tied to Jaeger and makes switching to a different implementation quite difficult.

Maybe my approach to use OT and Jaeger for this is not really suited for my use case. It would be great to hear your opinions on this :) I'm open to suggestions. Then again, the approach might not be wrong but I should accept that changing the OT implementation is going to be difficult because I need a quite deep integration. It's more important that the API for stagemonitor users is stable, should they want to add custom tags or logs to the spans.

I also see the downsides of adding the getters to the API. It makes implementations of the API less flexible, OT bridges more difficult and possibly more inefficient.

**Comment by yurishkuro** [comment]
@felixbarny the criteria you mention in the bullet points are quite interesting, we're also thinking of doing similar things, would love to collaborate.

As far as OT impact, your bullet points seem to fall into two categories:
  1. Affecting the tracer behavior. There is one precedent for this in the OTAPI, the `sampling.priority` tag that _may_ affect how tracer performs the sampling. Note the "_may_" qualifier - because the impact already implies an implementation detail, tracers are not required to respect this tag. The OTAPI was primarily designed as user facing, trying to make no assumptions about the implementations, thus it has none of the knobs to affect their behavior and leaves those to the implementations themselves. There was once a discussion about a shared _implementation_ framework (like the abstract tracer in this repo), but there was little interest from the actual tracer authors to use such framework, so it didn't go far.
  2. Causing side effects to the main OT operations. This can generally be achieved without changing the API, but via the decorator pattern. I couldn't find the issue link just now, but there were discussions about a shared multiplexing / "tee" tracer that would dispatch all OT API calls to two or more underlying tracers. I think that matches much better to your needs if you need to do other things in response to tracing instrumentation calls that are not necessarily related to tracing. Did you see [Ben Sigelman's talk](https://www.youtube.com/watch?v=M3sMM6tBbOY&t=https://youtu.be/M3sMM6tBbOY?t=21m51s) about using tracing instrumentation to emit metrics? In short, you create a decorator implementation of OT API that delegates all the calls to some real implementation (i.e. if you don't want to build a distributed tracing system yourself), but generates side effects suitable to your use case. For example, if you want to know about some tags, you can intercept the setTag calls and store some info in your own decorator span, where you have getters if you need it later.

**Comment by felixbarny** [comment]
@yurishkuro thx for the response and the link to Ben's talk! It was quite interesting to me because this is essentially what stagemonitor aims to do - combining tracing and metrics. It already does that but the thing thats missing yet is distributed tracing. I would love to collaborate on this as well :)

Using the decorator pattern seems like a good an reasonable idea to be independent of the implementation. It however comes with the downside of a bit of additional overhead for example by having two Maps which include the exact same tags.

I also thought about implementing my own Span which is actually not a lot of work. The thing that lets me shy away from it is that I'd also have to implement my own zipkin reporter. Tbh, I don't want to reimplement that again and maintain it.

This morning I had an idea about introducing different "levels" for the OT API to have optional getters. Level 1 implementations wouldn't support the getters and would throw an `UnsupportedOperationException`, similar to `java.util.Iterator#remove`. Level 2 implementation support the getters. That would enable something quite powerful: There could be a `/opentracing/opentracing-java-zipkin-reporter` project which would work with each Level 2 compliant implementation of the OT API out of the box. Cutting down a lot of copy and paste work in the tracing community and making it easier to implement the API, as the reporters can be reused. Also, 3rd party reporters could be reused in each Level 2 implementation. For example, I have implemented a Elasticsearch reporter which converts the Spans into a Kibana friendly JSON document.

Regarding the tracer behavior, stagemonitor offers two SPIs which lets clients intercept before and after the execution and let them decide whether or not to report the span. Stagemonitor also uses these SPIs internally, for example to [Exclude external requests from reporting on x% of the fastest external requests](https://github.com/stagemonitor/stagemonitor/wiki/Configuration-Options#exclude-external-requests-from-reporting-on-x-of-the-fastest-external-requests). I still have to figure out how to address `sampling.priority` in stagemonitor.

**Comment by yurishkuro** [comment]
> It however comes with the downside of a bit of additional overhead for example by having two Maps which include the exact same tags.

My assumption was that the decorator (metrics) tracer would not need to store the tags in the map. It would probably only be interested in some tags, and can look for them in real time and update some fixed fields of the wrapper span. This is the same approach as our Jaeger tracers do some special handling of the tags, like `span.kind` and `sampling.priority` - those tags are never even stored in the map, they just cause side effects and change fixed fields of the Span object.

> The thing that lets me shy away from it is that I'd also have to implement my own zipkin reporter. Tbh, I don't want to reimplement that again and maintain it.

I am not sure why you need that. With decorator tracer you would be using one real tracer that already has a tracing reporter (for example, Jaeger Java tracer can report to Zipkin). And the decorator tracer would only deal with metrics. That decorator tracer wouldn't even have any reporter, because metrics would normally be reporter via their own infrastructure.

My team has a metrics-emitting tracer in our OKRs for this quarter, so I expect we'd write one sometime in Feb as a POC (most likely in Go first).

> Also, 3rd party reporters could be reused in each Level 2 implementation.

There has been criticism of OpenTracing that it does not define in-band propagation or out-of-band reporting formats. That was a very deliberate decision, because it would significantly expand the scope of OT and make the spec much harder to agree on, given that even the concepts of trace/span ID can (and do) differ in different tracer implementations. I think in its current form the OT has solved a massive problem in the industry by providing portable instrumentation API. And now there's increased interest in solving these format problems, but it's hard without imposing on implementation how they should represent the data.

**Comment by felixbarny** [comment]
In my current architecture, tracking of the metrics is performed inside span reporters but I kinda like the decorator approach. Fells like a clean separation of concerns. I'll try it and see if it works out for me. Thx a bunch for the suggestion.

> There has been criticism of OpenTracing that it does not define in-band propagation or out-of-band reporting formats.

I totally agree that OT should not define propagation formats, that was not my point. But with included getters, the community could share reporters which work independently of the OT implementation. So that OT implementations don't need to implement their own reporters but can leverage existing work, not needing to "re-implement the wheel".

**Comment by felixbarny** [comment]
Just want to share a few details of my implementation:

One problem I'm facing with the wrapper approach is that tags can not only be set on `Span` but also on `SpanBuilder`. To solve this, I created a `SpanInterceptor` which has `onSetTag` methods which are called both in the my wrapper classes `SpanWrapper` and in `SpanWrappingSpanBuilder`. `SpanInterceptor`also contains the methods `onSetOperationName(String operationName)`, `onSetStartTimestamp(long startTimestamp)` and `onFinish(long endTimestamp)`.

As there is no `getDuration`, the start and end timestamps have to generated for the interceptor again, which is a bit unfortunate.

**Comment by yurishkuro** [comment]
@felixbarny can you share your interceptor code?

**Comment by felixbarny** [comment]
Sure. I'm still tweaking my implementation but I'll notify you as soon as I push.

**Comment by felixbarny** [comment]
This is my first shot: https://github.com/stagemonitor/stagemonitor/commit/fa50285736134604e0b9d8de60b24e3ae23d5154

I'd also like to be able to manipulate the tags via the interceptor. For example to anonymize sensitive information like the username or the IP. Not sure yet how to do that.

**Comment by yurishkuro** [comment]
Thanks for the link. The approach looks reasonable to me.

**Comment by michaelsembwever** [comment]
I have no objection to adding `span.getTags()` and `span.getTag(name)` and `getDuration()`
The interceptor code demonstrates the example.

**Comment by bensigelman** [comment]
I'm pretty resistant to Span getters with the exception of baggage- and SpanContext-related methods... since many implementations rely on sampling for performance reasons, noop Span implementations should be permitted to, well, be noops. (Which means that the getters would not return real data)

@felixbarny I am sympathetic about the desire for well-factored SPI-like helpers (which actually exist in OT-Java, at least to an extent)... though I wouldn't want to create helpers that are *required* for OpenTracing implementations.

Also, I'd be delighted to talk about metrics+tracing (or whatever else) if you want, though that is off-topic :)

**Comment by felixbarny** [comment]
> noop Span implementations should be permitted to, well, be noops. (Which means that the getters would not return real data)

That argument indeed makes this proposal obsolete. So I guess the wrapper/interceptor approach is the only way to go.

**Comment by felixbarny** [comment]
Just FYI: I successfully migrated the code to the span interceptor approach. Now the only references to com.uber.jaeger.Span are in custom Span reporters (logging reporter and Elasticsearch reporter). I have also changed SpanInterceptor a bit so that interceptors can modify the tag value. Also note the arguments of onFinish.

```java
String onSetTag(String key, String value);
boolean onSetTag(String key, boolean value);
Number onSetTag(String key, Number value);
void onFinish(Span span, String operationName, long durationNanos);
```

See also https://github.com/stagemonitor/stagemonitor/tree/distributed-tracing/stagemonitor-requestmonitor/src/main/java/org/stagemonitor/requestmonitor/tracing/wrapper
